addons/mrp/models/mrp_production.py

```diff
@@ -2073,6 +2120,135 @@ def action_merge(self):
     def _has_workorders(self):
         return self.workorder_ids
 
+    def _clean_moves_and_operations_and_assign_bom(self, bom_id):
+        self.ensure_one()
+        if self.state not in ['draft', 'confirm']:
+            return
+        if self.state != 'confirm':
+            self.write({'bom_id': bom_id})
+            bom = self.env['mrp.bom'].browse(bom_id)
+            for i, bom_line in enumerate(bom.bom_line_ids):
+                self.move_raw_ids[i].bom_line_id = bom_line.id
+            for i, byproduct in enumerate(bom.byproduct_ids):
+                self.move_byproduct_ids[i].byproduct_id = byproduct.id
+            for i, operation in enumerate(bom.operation_ids):
+                self.workorder_ids[i].operation_id = operation.id
+            return
+        moves_to_unlink = (self.move_finished_ids | self.move_raw_ids)
+        workorders_to_unlink = self.workorder_ids
+        self.write({'bom_id': bom_id})
+        # Unlinks all moves and workorders as they will be recreated linked to the BoM.
+        moves_to_unlink._action_cancel()
+        moves_to_unlink.unlink()
+        workorders_to_unlink.unlink()
+
+    def _link_bom(self, bom):
+        """ Links the given BoM to the MO, assigning the BoM's lines, by-products and operations
+        to matching MO's components, by-products and workorders.
+        This method assumes the production isn't linked to any BoM yet.
+        """
+        self.ensure_one()
+        moves_to_unlink = self.env['stock.move']
+        workorders_to_unlink = self.env['mrp.workorder']
+        # For draft MO, all the work will be done by compute methods.
+        # For cancelled and done MO, we don't want to do anything more than assinging the BoM.
+        if self.state == 'draft' and self.bom_id == bom:
+            self.bom_id = False
+        if self.state in ['cancel', 'done', 'draft']:
+            if self.state == 'draft':
+                moves_to_unlink = (self.move_raw_ids + self.move_finished_ids)
+                workorders_to_unlink = self.workorder_ids
```

Comment:
If I delete them immediatly, the MO is then cancelled.
By changing the BoM first, the new moves are created, so delete the old ones will not cancel the MO since it will still have some move.